vm-setup/roles/v1aX_integration_test/tasks/node_reuse.yml

```diff
@@ -182,175 +223,247 @@
-    shell: |
-        kubectl scale machinedeployment "{{ CLUSTER_NAME }}" -n "{{ NAMESPACE }}" --replicas=1
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    k8s:
+      api_version: cluster.x-k8s.io/{{ CAPI_VERSION }}
+      kind: MachineDeployment
+      name: "{{ CLUSTER_NAME }}"
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+      definition:
+        spec:
+          replicas: 1
 
   - name: Wait until "{{ NUMBER_OF_BMH_MD }}" more BMH is provisioned.
-    shell: kubectl get bmh -n "{{ NAMESPACE }}" | grep -w provisioned | wc -l
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: bmhs
     retries: 200
     delay: 20
-    register: provisioned_host
-    until: provisioned_host.stdout == "2"
+    until:
+      - bmhs is succeeded
+      - bmhs.resources | filter_provisioning("provisioned") | length == 2
 
   - name: Wait until "{{ NUMBER_OF_BMH_MD }}" more machine becomes running.
-    shell: |
-        kubectl get machines -n "{{ NAMESPACE }}" -o json | jq -r '.items[] | select (.status.phase == "Running") | .status.phase' | grep -c "Running"
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: provisioned_machine
+    k8s_info:
+      api_version: cluster.x-k8s.io/{{ CAPI_VERSION }}
+      kind: machines
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: machines
     retries: 200
     delay: 20
-    until: provisioned_machine.stdout|int == 2
+    until:
+      - machines is succeeded
+      - machines.resources | filter_phase("running") | length == 2
+
+  - name: Get provisioned BMH before upgrade in MachineDeployment node reuse test scenario.
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: saved_bmhs
 
   - name: Get provisioned BMH names and UUIDs mapping before upgrade in MachineDeployment node reuse test scenario.
-    shell: |
-        kubectl get bmh -A -o json | jq '.items[]| select (.status.provisioning.state == "provisioned")
-        | "metal3/"+.metadata.name+"="+"metal3://"+.metadata.uid' | cut -f2 -d\" | sort > /tmp/before_upgrade_mapping_md.txt
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    set_fact:
+      bmh_before_upgrade_md: "{{ bmh_before_upgrade_md | default([]) + [item.metadata.name + '=metal3://' + item.metadata.uid] }}"
+    with_items: "{{ saved_bmhs.resources | filter_provisioning('provisioned') }}"
 
   - name: Update maxSurge and maxUnavailable fields in MachineDeployment.
```

Comment:
I think it is better to explicitly define the values of `maxSurge` and `maxUnavailable` here too as you are doing in other places?